readthedocs/vcs_support/backends/git.py

```diff
@@ -185,7 +195,34 @@ def checkout_revision(self, revision):
 
     def clone(self):
         """Clones the repository."""
-        cmd = ['git', 'clone', '--no-single-branch']
+        cmd = ["git", "clone"]
+
+        # If we are building a branch, we can use self.version_identifier to tell Git to clone
+        # this branch history.
+        if self.use_clone_single_branch():
+            if self.version_type == BRANCH and self.version_identifier:
+                cmd.extend(["--branch", self.version_identifier])
+
+            # We could also enable this behavior for tags as shown below
+            # elif self.version_type == TAG and self.version_identifier:
+            #    cmd.extend(["--branch", self.version_identifier])
+
+            # Legacy behavior (this is intended for tags)
+            # If we cannot specify the branch to clone, we should get history for all remote types
+            # External builds aren't relevant, there is always an individual fetch operation.
+            # See self.update()
+            elif not self.version_type == EXTERNAL:
+                cmd.extend(["--no-single-branch"])
```

Comment:
I'd like to always have the same behavior for branch, tags and external versions. Otherwise, things are going to be pretty complicated to debug. Also, there will be things that work on PRs but fail on regular versions --which is the main feature of PRs builds.

Reply:
Let's go over this tomorrow then :+1: If you can set aside some time to read the initial analysis in https://github.com/readthedocs/readthedocs.org/issues/9736#issuecomment-1587373521, that should give the pretext of why this was necessary.